File: 프로그래머스/1/92334. 신고 결과 받기/신고 결과 받기.py
def solution(id_list, report, k):
    answer = [0] * len(id_list)
    report_list = {}
    report_count = {}
    for i in id_list :
        report_count[i] = 0
        report_list[i] = []
    
    for i in report :
        temp = i.split(" ")
        if temp[1] not in report_list[temp[0]] :
            report_list[temp[0]].append(temp[1])
            report_count[temp[1]] += 1
    
    temp = list(report_count.items())
    for i, j in temp :
        if j < k :
            report_count.pop(i)
    
    for i, j in enumerate(list(report_list.values())) :
        for k in j :
            if k in report_count :
                answer[i] += 1
    return answer

File: 프로그래머스/1/92334. 신고 결과 받기/신고 결과 받기.java
import java.util.*;

class Solution {
    public static int[] solution(String[] id_list, String[] report, int k) {
        Map<String, HashSet<String>> map = new HashMap<>();
        Map<String, Integer> index = new HashMap<>();
        int[] count = new int[id_list.length];
        HashSet<String> reportMember = new HashSet<>();

        for (int i = 0; i < id_list.length; i++) {
            index.put(id_list[i], i);
        }

        for (String r : report) {
            String[] split = r.split(" ");
            HashSet<String> value = map.getOrDefault(split[0], new HashSet<>());
            int before = value.size();
            value.add(split[1]);
            int after = value.size();
            map.put(split[0], value);

            if (before != after) {
                count[index.get(split[1])]++;
            }

            if (count[index.get(split[1])] >= k) {
                reportMember.add(split[1]);
            }
        }

        int[] answer = new int[id_list.length];
        for (int i = 0; i < id_list.length; i++) {
            HashSet<String> target = map.get(id_list[i]);
            if (target != null) {
                target.retainAll(reportMember);
                answer[index.get(id_list[i])] = target.size();
            }
        }
        return answer;
    }
}